### PR TITLE
feat(onyx-355): Apply filters on artist artworks grid when opened via deep link

### PR DESCRIPTION
**Pull request by nickskalkin**
This PR resolves [ONYX-355]

### Description

Filters are not applied when opening a link like [this one](https://staging.artsy.net/artist/kaws?sort=-partner_updated_at&attribution_class%5B0%5D=unique&additional_gene_ids%5B0%5D=prints&price_range=2000-49100) in the app. This PR addresses the issue:

https://github.com/artsy/eigen/assets/3934579/4b50552c-4aaf-41b2-8083-d7efce9677de

> [!NOTE]  
> Current PR addresses the issue only for artist's artworks screen.

There are other grids with filters, where the problem still remains. I think we should not fix all of them in the same PR, but introduce changes gradually in separate PRs. Here is a list of other grids that require changes:

| Name | URL |
|---|---|
| Artist Series | https://staging.artsy.net/artist-series/kaws-4-foot-companion?attribution_class%5B0%5D=unique&additional_gene_ids%5B0%5D=sculpture |
| Auction | https://staging.artsy.net/auction/heritage-urban-art-53?sort=sale_position&attribution_class%5B0%5D=unknown%20edition&attribution_class%5B1%5D=open%20edition&additional_gene_ids%5B0%5D=prints |
| Artist Auction Results | https://staging.artsy.net/artist/banksy/auction-results?categories%5B0%5D=Painting&hide_upcoming=false&allow_empty_created_dates=true&currency=&include_estimate_range=false&include_unknown_prices=true&allow_unspecified_sale_dates=true |
| Collection | https://staging.artsy.net/collection/photography?attribution_class%5B0%5D=unique&additional_gene_ids%5B0%5D=photography |
| Gene | https://staging.artsy.net/gene/photography?attribution_class%5B0%5D=unique&additional_gene_ids%5B0%5D=photography |
| Gallery | https://staging.artsy.net/partner/kumar-gallery/works?attribution_class%5B0%5D=unique&additional_gene_ids%5B0%5D=painting |
| Fair | https://staging.artsy.net/fair/the-art-show-2023/artworks?attribution_class%5B0%5D=unique&additional_gene_ids%5B0%5D=painting |
| Show | https://staging.artsy.net/show/berry-campbell-gallery-larry-zox-gemini?sort=partner_show_position&attribution_class%5B0%5D=unique&additional_gene_ids%5B0%5D=painting |
| Tag | https://staging.artsy.net/tag/easter?attribution_class%5B0%5D=unique |

### PR Checklist

- [x] I have tested my changes on **iOS**.
- [ ] I have tested my changes on **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [x] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

- feat: apply filters on artist artworks grid when opened via deep link - @nickskalkin 

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md

[ONYX-355]: https://artsyproduct.atlassian.net/browse/ONYX-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
### This PR contains the following changes:

- Cross-platform user-facing changes (bugfix: fixes filters are not applied when opening an artist's artworks deep link with filters query params - @nickskalkin)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 2c6cb5c5b4751b40d3f557cd63cd0c7c68b69083
</p>

**Comment by evaschilken** [comment]
*Nudge to not let this PR die in oblivion*

File: src/app/Components/ArtworkFilter/Filters/SortOptions.tsx [hunk 1]
```diff
@@ -47,12 +47,12 @@ export const ORDERED_ARTWORK_SORTS: FilterData[] = [
   {
     displayText: "Price (High to Low)",
     paramName: FilterParamName.sort,
-    paramValue: "sold,-has_price,-prices",
```
Comment (nickskalkin) [inline src/app/Components/ArtworkFilter/Filters/SortOptions.tsx:50]: I think it is safe to drop `sold` here. I've made requests to MP with and without it and results are the same

File: src/app/Scenes/Artist/Artist.tsx [hunk 2]
```diff
@@ -234,12 +247,19 @@ export const ArtistQueryRenderer: React.FC<ArtistQueryRendererProps> = (props) =
...
           let initialArtworksInput = {
             ...defaultArtistVariables().input,
             ...predefinedFilterParams,
           }
 
+          const allowedFilterRouteParams = allowedFilterParams(routeParams)
+
+          if (allowedFilterRouteParams?.sort) {
+            initialArtworksInput.sort = allowedFilterRouteParams.sort
+          }
```
Comment (MounirDhahri) [inline src/app/Scenes/Artist/Artist.tsx:261]: is there something special about sort that we would like to handle differently?

**Review by MounirDhahri — APPROVED** [review]
LGTM! thanks for adding the mapping as well, this will make it a lot more future proof

File: src/app/Scenes/Artist/Artist.tsx (hunk 2 above)
Comment (nickskalkin) [reply src/app/Scenes/Artist/Artist.tsx]: I think this is redundant, thanks for noticing!

**Review by MounirDhahri — COMMENTED** [review]
Thanks for the fix, sorry I noticed something I missed the first time.

File: src/app/Components/ArtworkFilter/ArtworkFilterHelpers.ts [hunk 3]
```diff
@@ -77,6 +77,39 @@ export type FilterParams = {
...
 }
 
+export const QueryParamsToFilterValueMapping: Record<string, FilterParamName> = {
+  additional_gene_ids: FilterParamName.additionalGeneIDs,
+  artist_ids: FilterParamName.artistIDs,
+  artist_nationalities: FilterParamName.artistNationalities,
+  artist_series_ids: FilterParamName.artistSeriesIDs,
+  include_artworks_by_followed_artists: FilterParamName.artistsIFollow,
+  attribution_class: FilterParamName.attributionClass,
+  something: FilterParamName.categories,
```
[the remark this answers is not in the document]
Comment (nickskalkin) [reply src/app/Components/ArtworkFilter/ArtworkFilterHelpers.ts:87]: there is a mapping `additional_gene_ids: FilterParamName.additionalGeneIDs` already. I think we don't need one for `categories`

File: src/app/Components/ArtworkFilter/ArtworkFilterHelpers.ts [hunk 4]
```diff
@@ -584,3 +615,27 @@ export const getFilterArrayFromQueryParams = (queryParams: {
...
+
+export const getFilterParamsFromRouteParams = (routeParams: {
+  [key: string]: string | number | boolean | string[]
+}) => {
+  return Object.entries(allowedFilterParams(routeParams)).map(([key, value]) => {
+    const paramName = QueryParamsToFilterValueMapping[key]
+    let paramValue = value
+    if (paramValue === "true" || paramValue === "false") {
+      paramValue = paramValue === "true" ? true : false
+    }
```
Comment (olerichter00) [inline src/app/Components/ArtworkFilter/ArtworkFilterHelpers.ts:633]: Maybe this is easier to read 🤷 

```suggestion

    let paramValue = value === true || value === "true"
```

File: src/app/Scenes/Artist/Artist.tsx [hunk 5]
```diff
@@ -233,6 +235,16 @@ export const ArtistQueryRenderer: React.FC<ArtistQueryRendererProps> = (props) =
     sizes,
   } = props
 
+  // exctact filter params from the query string. This is needed when
+  // the screen is opened via deeplink (/artist/kaws?attribution_class=..., for instance)
+  // to make sure the filters are applied correctly
+  const route = useRoute()
+  const routeParams = (route?.params as any)?.props || {}
```
Comment (olerichter00) [inline src/app/Scenes/Artist/Artist.tsx:242]: Do you think we could get rid of the `as any` somehow?

**Review by olerichter00 — DISMISSED** [review]
Looks good 🌟

File: src/app/Scenes/Artist/Artist.tsx (hunk 5 above)
Comment (nickskalkin) [reply src/app/Scenes/Artist/Artist.tsx:242]: I can create an interface like this, but not sure if it worth it, wdyt?

```
interface RouteParams {
  props: {
    [key: string]: any; // Replace 'any' with the actual type of the props
  };
}
```

Comment (nickskalkin) [reply src/app/Scenes/Artist/Artist.tsx]: added it @olerichter00

File: src/app/Components/ArtworkFilter/ArtworkFilterHelpers.ts (hunk 4 above)
Comment (olerichter00) [reply src/app/Components/ArtworkFilter/ArtworkFilterHelpers.ts:633]: Ah, ok. What would happen if it is "foobar"?

Comment (nickskalkin) [reply src/app/Components/ArtworkFilter/ArtworkFilterHelpers.ts:633]: nothing should happen - `paramValue` should have the value of `value`

**Pull request merged**